### PR TITLE
Store snapshots objects count in snapshot_obj_1

### LINKED ISSUE #363: Save the file count information in the snapshot to ensure it is a complete snapshot file and prevent data loss.
### RaftKeeper version
v2.1.1
### Description
Found in test https://github.com/JDRaftKeeper/RaftKeeper/actions/runs/11028685908/job/30630210401?pr=361
We kill RaftKeeper force, only file `/snapshot/snapshot_20240925082106_1_10_1` created
```
2024-09-25T08:23:54.1793916Z 2024.09.25 08:21:06.528459 [ 12 ] <Information> KeeperSnapshotStore: Creating async snapshot v3 with approximately data_object_count 1, total_obj_count 4, next zxid 9, next session id 3
2024-09-25T08:23:54.1794582Z 2024.09.25 08:21:06.528524 [ 12 ] <Information> KeeperSnapshotStore: Begin create snapshot map object, map size 2, path /snapshot/snapshot_20240925082106_1_10_1
```
After that, we restart RaftKeeper, and loss all data
```
2024-09-25T08:23:54.1804421Z 2024.09.25 08:21:06.805812 [ 178 ] <Information> KeeperSnapshotStore.parseObjectThread#0: Parsing snapshot object /snapshot/snapshot_20240925082106_1_10_1
2024-09-25T08:23:54.1804963Z 2024.09.25 08:21:06.806267 [ 178 ] <Information> KeeperSnapshotStore: Open snapshot object /snapshot/snapshot_20240925082106_1_10_1 for read, file size 82
2024-09-25T08:23:54.1805298Z 2024.09.25 08:21:06.806400 [ 171 ] <Information> KeeperSnapshotStore: Parsing snapshot objects costs 2ms
2024-09-25T08:23:54.1805669Z 2024.09.25 08:21:06.806464 [ 171 ] <Information> KeeperSnapshotStore: Building data tree from snapshot objects
...
2024-09-25T08:23:54.1821628Z 2024.09.25 08:21:06.808007 [ 171 ] <Information> KeeperSnapshotStore: Loading snapshot done: nodes 1, ephemeral nodes 0, sessions 0, session_id_counter 3, zxid 9
```

### PR DESCRIPTION
**Pull request by lzydmxy**
### Which issues of this PR fixes:

This PR try to fix #363

### Change log:
Store snapshots objects count in snapshot_obj_1

### CODE REVIEW

File: src/Service/NuRaftLogSnapshot.cpp [hunk 1]
```diff
@@ -555,7 +561,17 @@ void KeeperSnapshotStore::parseBatchBodyV2(
...
 {
-    auto objects_cnt = objects_path.size();
+    size_t objects_cnt = objects_path.size();
+
+    // The object IDs are consecutive starting from 1,
+    // so the first number must be 1, and the last number must be the total count.
+    if (objects_path.begin()->first != 1 || objects_path.rbegin()->first != objects_cnt)
+    {
+        throw Exception(ErrorCodes::SNAPSHOT_OBJECT_INCOMPLETE,
+        "Load snapshot objects error, except 1 ~ {} objects, got {} ~ {} objects",
```
Comment (JackyWoo) [inline src/Service/NuRaftLogSnapshot.cpp:571]: Load -> Loading, except -> expecting

File: src/Service/NuRaftLogSnapshot.h [hunk 2]
```diff
@@ -258,6 +260,9 @@ class KeeperSnapshotStore
 
     std::map<ulong, String> objects_path;
 
+    /// Except snapshot object counts, read from object1
+    std::optional<UInt32> load_objects_count;
```
[the remark this answers is not in the document]
Comment (JackyWoo) [reply src/Service/NuRaftLogSnapshot.h:264]: We also should update load_objects_count after creating snapshot.

File: src/Service/NuRaftLogSnapshot.cpp [hunk 3]
```diff
@@ -586,6 +602,13 @@ void KeeperSnapshotStore::loadLatestSnapshot(KeeperStore & store)
     thread_pool.wait();
     LOG_INFO(log, "Parsing snapshot objects costs {}ms", watch.elapsedMilliseconds());
 
+    if (load_objects_count && *load_objects_count != objects_path.size())
```
Comment (JackyWoo) [inline src/Service/NuRaftLogSnapshot.cpp:605]: Maybe it is better to move the checking in the front of loading data tree. We can load object1 first and then load data tree in parellel. And we can do it latter when squashing the first 3 objects.

Comment (lzydmxy) [reply src/Service/NuRaftLogSnapshot.cpp:605]: Checking load_objects_count here is an ugly implementation, but it avoids some code changes and refactoring. I will do it  when squashing the first 3 objects.

File: src/Service/NuRaftLogSnapshot.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (lzydmxy) [reply src/Service/NuRaftLogSnapshot.h:264]: The loaded_objects_count is only relevant when loading a snapshot. Perhaps we don't need to update it after creating the snapshot.

[the remark this answers is not in the document]
Comment (JackyWoo) [reply src/Service/NuRaftLogSnapshot.h:264]: ok

**Review by JackyWoo — APPROVED** [review]

**Pull request merged**